Add LIP "Introduce cross-chain update mechanism" (#106)
LiskHQ/lips: proposals/lip-Introduce_cross-chain_update_mechanism.md

#### certificate

Used to update the cross-chain account. It must be a certificate with an adequate [BLS signature][BLS-LIP]. 
The signature is always validated against the public key set stored in the sending chain account.

It is important to note here that the validation of this signature is done with respect to the network identifier of the sending chain. 
This network identifier is included in the sending chain account in the interoperability store.

#### activeValidatorsUpdate

The chain account stores an array containing the bls keys and bft weights required to validate the certificate signature. 
This array has to be updated if the validator set changes in the chain sending the certificate. 
The difference between the stored array and the new one is included in this property.

#### newCertificateThreshold

The chain account stores the threshold required to validate the certificate signature. 
This certificate threshold has to be updated if its value changes in the chain sending the certificate,
In that case, the updated value is set in the `newCertificateThreshold` property of the CCU.

> 

Reviewer: Maybe here we could clarify: "If `newCertificateThreshold` is set to 0, it implies that the certificate threshold has not been changed in the sending chain.

Author: nice one, added.

Those options for partial cross-chain update transactions are not expected to be used regularly. 
It is however a good option in corner case events where the sidechain outbox suddenly becomes very large.

#### Malicious Cross-chain Update Transactions

A cross-chain update transaction may have been created with validly signed information according to the [validity rules](#parameters-validity) and hence be included in the blockchain. 
However, the transaction could include invalid CCMs. 
Those will be detected when trying to process the `inboxUpdate`. 
If a CCM is invalid (as specified in the ["Execute Cross-chain Updates" section](#execute-cross-chain-updates)), then the sending chain is [terminated][base-interoperability-LIP-termination].

#### First Cross-chain Update from a Sidechain

The first cross-chain update containing messages from a given chain has a special function. 
It will change the sending chain status from `CHAIN_REGISTERED` to `CHAIN_ACTIVE`.
This change means that the receiving chain is now available to receive cross-chain messages and can interact with the sending chain.

> Additionally, once active, sidechains must now follow the liveness condition and regularly post cross-chain updates on the mainchain (at least once a month).

Reviewer: ```suggestion
Additionally, once active, sidechains must follow the liveness condition and regularly post cross-chain updates on the mainchain (at least once every 30 days).
```

| **Chain Status** |||
| `CHAIN_REGISTERED` | uint32 | 0 |
| `CHAIN_ACTIVE`     | uint32 | 1 |
| `CHAIN_TERMINATED` | uint32 | 2 |
| **Message Tags** |||
| `MESSAGE_TAG_CERTIFICATE ` | bytes | "LSK_CE_" ASCII-encoded |

Several of those constants are shared with the other LIPs defining the Interoperability module and all of the needed constants for the Interoperability module are defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP]. 
That LIP should be considered correct if a value stated here differs.

#### uint32be

`uint32be(x)` returns the big endian uint32 serialization of an integer `x`, with `0 <= x < 2^32`. 
This serialization is always 4 bytes long.

### Parameters Schema

Both commands will use the following `params` schema:

```java
crossChainUpdateTransactionParams = {
    "type": "object",
    "required": [
        "sendingChainID", 
        "certificate", 
        "validatorsUpdate",
        "inboxUpdate"
    ],
    "properties": {
        "sendingChainID": {
            "dataType": "uint32",
            "fieldNumber": 1
        },
        "certificate": {
            "dataType": "bytes",
            "fieldNumber": 2
        },	

>         "validatorsUpdate": {

Reviewer: Just a reminder: If we update the chain account schema in the main interoperability LIP, it would be better to update this schema as well, and remove the nesting for the  "activeValidatorsUpdate", "newCertificateThreshold" properties

Author: yes, changed.

                            "fieldNumber": 1
                        },
                        "siblingHashes":{
                            "type": "array",
                            "fieldNumber": 2,
                            "items": {"dataType": "bytes"}
			}
                    }
                }
            }
        }
    }
}
```

### Parameters Validity

In the following, let `sendingAccount` be the entry in the interoperability store with store prefix `STORE_PREFIX_CHAIN_DATA` and store key `uint32be(sendingChainID)`.

#### Liveness of Sending Chain

A CCU transaction is only valid if the sending chain is not terminated and follows the [liveness rule][base-interoperability-LIP-livenessCondition]. 
This is done by asserting the two points below:  

*  `sendingChain.status != CHAIN_TERMINATED`,
*  if `sendingChain.status == CHAIN_ACTIVE` then also validate  `isLive(sendingChainID,timestamp) == True`. 
   The `isLive` function is specified in [LIP "Introduce Interoperability module"][base-interoperability-LIP], and `timestamp` is the timestamp of the block including the CCU.

##### Liveness Requirement for the First CCU

> If `sendingChain.status == CHAIN_REGISTERED`, the proposed CCU must contain a non-empty `certificate` which must follow the schema defined in  [LIP "Introduce a certificate generation mechanism"][certificate-generation-LIP]. 

Reviewer: Doesn't this introduce a dependency on that LIP?

Reviewer: To follow up, how do we deserialize the certificate in practice? With an API call?

Author: I added the LIP as required.
For the follow up, I don't know how various schemas will be shared between modules. I would prefer to keep the language general (as is) but if you want to do it with an API call, we can probably add it.

In the following, let `certificate` be the deserialized certificate.

Furthermore, the certificate is only valid if it allows the sidechain account to remain live for a reasonable amount of time.
This is done by checking that
```java

> timestamp - CCU.params.certificate.timestamp < 15*24*3600

Reviewer: In the interoperability LIP I introduced the constant `LIVENESS_LIMIT` for this. We could add it here to the table as well

Reviewer: Also maybe use python formatting

Author: changed formatting.
changed to use `LIVENESS_LIMIT`.

```
where `timestamp` is the timestamp of the block including the CCU.

#### Certificate and Validators Update Validity

If `params` contains a non-empty `certificate`, it is valid if:

*  `params.certificate` follows the schema defined in [LIP "Introduce a certificate generation mechanism"][certificate-generation-LIP]. 
   Again, we write `certificate` for the deserialized certificate.
*  `certificate.height` is greater than `sendingAccount.lastCertifiedHeight`.
*  `certificate.timestamp < timestamp`, where timestamp is the timestamp of the block including the CCU.
*  `certificate.signature` is a valid aggregate signature for the `sendingAccount.validators` property of the cross-chain account. This is done by verifying that the function below defined in the [LIP 0038][BLS-LIP-verifyWeightedAggSig] returns `VALID`. 	

> 	```java

Reviewer: Maybe use python formatting

Author: done

	verifyWeightedAggSig( 
	    keysList          = sendingAccount.validators.keys,  
	    aggregationBits.  = certificate.aggregationBits,  
	    signature         = certificate.signature,  
	    tag               = MESSAGE_TAG_CERTIFICATE, 
	    networkIdentifier = sendingAccount.networkID,  
	    weights           = sendingAccount.validators.weights,  
	    threshold         = sendingAccount.validators.threshold, 
	    message           = params.certificate 
	)
	```

If `params` contains a non-empty `validatorsUpdate` property (with deserialized value not equal to `{keysUpdate:[], weightsUpdate:[], newCertificateThreshold: 0}`), 
it is valid if:

*   `params` contains a non-empty `certificate`.
*   `validatorsUpdate` has the correct format:
    *   for all `activeValidator` in `validatorsUpdate.activeValidators` we have that `activeValidator.blsKey` is a BLS public keys, hence are 48 bytes long.
    *   every `activeValidator` in `validatorsUpdate.activeValidators` has a different `activeValidator.blsKey`.

> *   `certificate.validatorsHash` is obtained as the SHA-256 digest of the updated `sendingAccount.validators`, see ["Update Validators" section](#update-validators) below, serialized according to `chainAccountSchema.validators` defined in the [LIP "Introduce Interoperability module"][base-interoperability-LIP] (and copied below).

Reviewer: I think the idea could be to instead call the API from the BFT module (just a reminder, this is still not decided)

Author: done. Will update if name changes.

                }
            },
            "certificateThreshold": {
                "dataType": "uint64",
                "fieldNumber": 2
            }
        }
    }
    ```

#### InboxUpdate Validity

If `params` contains a non-empty `certificate` and an `inboxUpdate`, the validity of the latter is checked by:

*   Let `newInboxRoot`, `newInboxAppendPath` and `newInboxSize` be the resulting new root, append path and size of the Merkle tree that would be obtained if all messages from `crossChainMessages` would be appended to the `sendingAccount` inbox. 
    Note that `sendingAccount.inbox` is not updated here.
*   If `inboxUpdate` contains a non-empty `messageWitness`, then update `newInboxRoot` to the output of `calculateRootFromRightWitness(newInboxSize, newInboxAppendPath, inboxUpdate.messageWitness.siblingHashes)` as specified in [LIP 0031](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0031.md).
*   Then validate the newly updated root against the certificate state root using the provided `outboxRootWitness`. 
    Using notation from [LIP 0039][SMT-LIP], this is done via the function `verify(queryKeys, proof, certificate.stateRoot)` with 

>     ```java

Reviewer: maybe use python formatting

Author: daon

*   If `inboxUpdate` contains a `messageWitness`, then update `newInboxRoot` to the output of `calculateRootFromRightWitness(newInboxSize, newInboxAppendPath, inboxUpdate.messageWitness.siblingHashes)` as specified in [LIP 0031](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0031.md).
*   Then validate that `newInboxRoot == sendingAccount.partnerChainOutboxRoot`.

### Execute Cross-chain Updates

Cross-chain updates posted on the mainchain or sidechains follow a different execution logic, which we describe in the following two subsections, and a common one, described in the ["Common Processing" section](#common-processing). 

In the following, let `CCU` be the cross-chain update transaction. Let `partnerChainID = CCU.sendingChainID` and `partnerChain` the entry in the interoperability store with store prefix `STORE_PREFIX_ACCOUNT` and store key `uint32be(partnerChainID)`.

#### Cross-chain Updates Posted on Mainchain

Cross-chain update transactions posted on mainchain are transactions with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`,
*   `commandID  = COMMAND_ID_MAINCHAIN_CCU`. 

Then, the following is done in the given order:

> *   If `partnerChain.status == CHAIN_REGISTERED`, set `partnerChain.status = CHAIN_ACTIVE`

Reviewer: ```suggestion
*   If `partnerChain.status == CHAIN_REGISTERED`, set `partnerChain.status = CHAIN_ACTIVE`.
```

*   For every `CCM` in `inboxUpdate.crossChainMessages`:
    *   Validate that `params.sendingChainID == CCM.sendingChainID`.

>     *   Validate the format of `CCM` according to the function provided in [LIP "Introduce cross-chain messages"][CCM-LIP]. 

Reviewer: I think it's worth mentioning the function name and link to the function directly.
```suggestion
    *   Validate the format of `CCM` according to the `validateFormat` function provided in [LIP "Introduce cross-chain messages"](https://research.lisk.com/t/introduce-cross-chain-messages/299#validateformat-27). 
```

Reviewer: The same in the next section

Author: changed

*   For every `CCM` in `inboxUpdate.crossChainMessages` (respecting the order of the array):
    *   Call <code>[appendToInboxTree][base-interoperability-LIP-appendToInboxTree](partnerChainID, SHA-256(serializedMessage))</code> where `serializedMessage` is the serialized CCM according to the schema given in [LIP "Introduce cross-chain messages"][CCM-LIP].
    *   Process `CCM` as detailed in [LIP "Introduce cross-chain messages"][CCM-LIP-process].
*  Update `partnerChain.validators` according to `validatorsUpdate`, see ["Update Validators" section](#update-validators).
*  Set `partnerChain.lastCertifiedStateRoot` to `certificate.stateRoot`.
*  Set `partnerChain.lastCertifiedTimestamp` to `certificate.timestamp`.
*  Set `partnerChain.lastCertifiedHeight` to `certificate.height`.
*  Set `partnerChain.partnerChainOutboxSize` to
    *   `inboxUpdate.messageWitness.partnerChainOutboxSize` if `params` contains a non-empty `inboxUpdate.messageWitness`.
    *   `partnerChain.inbox.size` otherwise.
*   Set `partnerChain.partnerChainOutboxRoot` to

>     *   `calculateRootRightWitness(inbox.size, inbox.appendPath, inboxUpdate.messageWitness.siblingHashes)` if `params` contains a non-empty `inboxUpdate.messageWitness`.

Reviewer: The function `calculateRootRightWitness` here is undefined. Maybe we could add a link to the Merkle tree LIP.

Reviewer: Also the function name is `calculateRootFromRightWitness`

Author: changed name.
added link.

    *   `partnerChain.inbox.root` otherwise.
*   Append a cross-chain update receipt to the partner chain outbox by calling `addToOutobx(partnerChain, CCUR)` with

> 	```java

Reviewer: Maybe use python formatting

Author: done

This network identifier is included in the sending chain account in the interoperability store.

#### activeValidatorsUpdate

The chain account stores an array containing the bls keys and bft weights required to validate the certificate signature. 
This array has to be updated if the validator set changes in the chain sending the certificate. 
The difference between the stored array and the new one is included in this property.

> ### newCertificateThreshold

Reviewer: ```suggestion
#### newCertificateThreshold
```

Author: done

The chain account stores the threshold required to validate the certificate signature. 

> This certificate threshold has to be updated if its value changes in the chain sending the certificate,

Reviewer: ```suggestion
This certificate threshold has to be updated if its value changes in the chain sending the certificate.
```

    In this case, the `messageWitness` will be empty.
*   Compute the inclusion proof for the outbox root of the sidechain account into the mainchain state root. 
    This proof is then used to compute `inboxUpdate.outboxRootWitness`.
*   If the `validatorsHash` property of the certificate is different from the `validatorsHash` derived from the `activeValidators` and `certificateThreshold` properties of the mainchain account on the sidechain: 

>    *   Compute the required active validator update between the `activeValidators` stored in the chain account 

Reviewer: I find this sentence a bit confusing, maybe it could be simplified to smtg like:

"Compute the update of the `activeValidators` stored in the chain account by following the logic of `getActiveValidatorsDiff` as detailed in the [Appendix](#appendix). "

Author: yes, it was not the best wording. Rephrased.

	)
	```

If `params` contains a non-empty `activeValidatorsUpdate` property (with deserialized value not equal to `{activeValidatorsUpdate:[]}`), or a non-zero `newCertificateThreshold`, `params` is valid if:

*   `params` contains a non-empty `certificate`.
*   `params.activeValidatorsUpdate` has the correct format:

>     *   for all `activeValidator` in `params.activeValidatorsUpdate` we have that `activeValidator.blsKey` is a BLS public keys, hence are 48 bytes long.

Reviewer: ```suggestion
    *   for all `activeValidator` in `params.activeValidatorsUpdate` we have that `activeValidator.blsKey` is a BLS public keys, hence it is 48 bytes long.
```

*   For every `CCM` in `inboxUpdate.crossChainMessages` (respecting the order of the array):
    *   Call <code>[appendToInboxTree][base-interoperability-LIP-appendToInboxTree](partnerChainID, SHA-256(serializedMessage))</code> where `serializedMessage` is the serialized CCM according to the schema given in [LIP "Introduce cross-chain messages"][CCM-LIP].
    *   Process `CCM` as detailed in [LIP "Introduce cross-chain messages"][CCM-LIP-process].

> *  Set `partnerChain.activeValidators = updateActiveValidaros(partnerChain.activeValidators, activeValidatorsUpdate)` as specified in the ["Update Active Validators" section](#update-active-validators).

Reviewer: ```suggestion
*  Set `partnerChain.activeValidators = updateActiveValidators(partnerChain.activeValidators, activeValidatorsUpdate)` as specified in the ["Update Active Validators" section](#update-active-validators).
```

When posting a CCU transaction, the validators hash given in the certificate certifies the new set of validators of the sending chain. 
The CCU must therefore include the difference between the validators currently stored in the chain account and the validator set authenticated by the certificate.
The required `activeValidatorsUpdate` can be obtained by applying the function below.

> We suppose that `currentValidators` and `newValidators` follow the `activeValidators` schema given .

Reviewer: ```suggestion
We suppose that `currentValidators` and `newValidators` follow the `activeValidators` schema given above.
```
